### DISCUSSION TITLE
Define the order() argument from a variable

### DISCUSSION THREAD

**Discussion by agateau** [diesel-rs/diesel#2991] [Q&A] [answered]
I am new to Diesel, and I want to create a method returning a vector of models sorted by a method parameter. My current code looks like this:

```rust
pub enum Order {
    DATE,
    LENGTH,
    DURATION
}
// ...
    pub fn get_tracks(&self, order: Order) -> Vec<Track> {
        let order_arg = match order {
            DATE => {
                start_date_col.asc()
            },
            LENGTH => {
                length_col.asc()
            },
            DURATION => {
                duration_col.asc()
            }
        };

        tracks_table
            .order(order_arg)
            .load::<Track>(&self.conn)
            .unwrap()
    }
```

The `*_col` identifiers are my column names, which I use with an alias to avoid collisions, so `start_date_col` is `use crate::schema::tracks::columns::start_date as start_date_col`)

But this fails to compile because Rust says the matches are not the same type:

```
error[E0308]: `match` arms have incompatible types
  --> src/database.rs:82:17
   |
77 |           let order_arg = match order {
   |  _________________________-
78 | |             DATE => {
79 | |                 start_date_col.asc()
   | |                 -------------------- this is found to be of type `Asc<columns::start_date>`
80 | |             },
81 | |             LENGTH => {
82 | |                 length_col.asc()
   | |                 ^^^^^^^^^^^^^^^^ expected struct `columns::start_date`, found struct `columns::length`
...  |
86 | |             }
87 | |         };
   | |_________- `match` arms have incompatible types
   |
   = note: expected type `Asc<columns::start_date>`
            found struct `Asc<columns::length>`
```

**Comment by weiznich** [comment, marked as answer]
As the compiler message says each match arm returns a different type. That's expected as each column type is it's own distinct type. See the [Schema in depth](https://diesel.rs/guides/schema-in-depth.html) guide for details about what is produced by diesel's `table!` macro internally.

Now there are multiple ways to "solve" your problem:

* Construct the query before the `match` statement, use [`.into_boxed()`](https://docs.diesel.rs/1.4.x/diesel/query_dsl/trait.QueryDsl.html#method.into_boxed) to produce a stable query type and then apply the order in each match arm
* Use [`BoxableExpression`](https://docs.diesel.rs/1.4.x/diesel/expression/trait.BoxableExpression.html) to construct a compatible common trait object based on your expressions.

**Reply by agateau** [reply]
Using `.into_boxed()` works fine, thanks!
